File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Services/PanoUserService.m
//
//  PanoUserService.m
//  PanoVideoCall
//
//  Created by pano on 2020/9/28.
//  Copyright © 2020 Pano. All rights reserved.
//

#import "PanoUserService.h"
#import "PanoClientService.h"
#import "UIImage+Extension.h"

@interface PanoUserService ()
@property (nonatomic, strong) NSMutableArray<PanoUserInfo *> *dataSource;
@property (nonatomic, strong, readonly) NSMutableDictionary *audioLevels;
@property (nonatomic, strong) NSDate *lastSpeakTime;

@property (nonatomic, strong) NSMutableDictionary *avators;
@property (nonatomic, assign) NSInteger imagesCount;
@property (nonatomic, strong) NSMutableArray *images;
@property (nonatomic, strong) NSMutableArray *offlineUsers;
@end

@implementation PanoUserService

- (void)initService {
    _dataSource = [NSMutableArray array];
    _audioLevels = [NSMutableDictionary dictionary];
    _offlineUsers = [NSMutableArray array];
    
    // 随机头像
    _avators = [NSMutableDictionary dictionary];
    _imagesCount = 22;
    _images = [NSMutableArray arrayWithCapacity:_imagesCount];
    for (NSInteger i=0; i<_imagesCount; i++) {
        [_images addObject:@(false)];
    }
}

- (void)notifyRefresh:(PanoUserInfo *)user {
    if (!PanoClientService.service.joined) return;
    [self invokeWithAction:@selector(onUserPropertyChanged:) completion:^(id <PanoUserDelegate> _Nonnull del) {
        [del onUserPropertyChanged:user];
    }];
}

- (void)onUserUpdated:(UInt64)userId message:(NSDictionary *)info {
    PanoUserInfo *user = [self findUserWithId:userId];
    [user setValuesForKeysWithDictionary:info[@"palyload"]];
}

- (void)onUserJoinIndication:(UInt64)userId withName:(NSString * _Nullable)userName {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSLog(@"onUserJoinIndication: %lld, %@", userId, userName);
        if (!PanoClientService.service.joined) return;
        PanoUserInfo *user = [[PanoUserInfo alloc] initWithId:userId name:userName];
        [self.offlineUsers removeObject:user];
        if ([self.dataSource containsObject:user]) return;
        if (user.userId == PanoClientService.service.config.userId) {
            [self.dataSource insertObject:user atIndex:0];
        } else {
            [self.dataSource addObject:user];
        }
        [self invokeWithAction:@selector(onUserAdded:) completion:^(id <PanoUserDelegate>  _Nonnull del) {
            [del onUserAdded:user.copy];
        }];
    });
}

- (void)onUserLeaveIndication:(UInt64)userId
                   withReason:(PanoUserLeaveReason)reason {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (!PanoClientService.service.joined) return;
        PanoUserInfo *user = [self findUserWithId:userId];
        NSLog(@"onUserLeaveIndication: %@, %ld", user, (long)reason);
        if (reason == kPanoLeaveDisconnected) {
            if (![self.offlineUsers containsObject:user]) {
                [self.offlineUsers addObject:user.copy]; // 缓存重连的用户
            }
            [self performSelector:@selector(checkOfflineUser:) withObject:[user copy] afterDelay:5];
        } else {
            // 断线用户重连成功后其它原因离开
            [self.offlineUsers removeObject:user];
            [self onUserLeave:user];
        }
    });
}

- (void)checkOfflineUser:(PanoUserInfo *)user {
    /**
     1. 检查10秒钟是否重连成功，无论重连成功与否，都从offlineUsers删除掉。
     2. 如果offlineUsers 还包含当前用户，那么通知上层
     3. 如果offlineUsers 不包括当前用户，说明重连成功
     */
    if ([self.offlineUsers containsObject:user] ) {
        [self onUserLeave:user];
    }
    [self.offlineUsers removeObject:user];
}

- (void)onUserLeave:(PanoUserInfo *)user {
    [self.dataSource removeObject:user];
    [self invokeWithAction:@selector(onUserRemoved:) completion:^(id <PanoUserDelegate>  _Nonnull del) {
        [del onUserRemoved:[user copy]];
    }];
}

- (void)onUserAudioStart:(UInt64)userId {
    dispatch_async(dispatch_get_main_queue(), ^{
        PanoUserInfo *user = [self findUserWithId:userId];
        user.audioStatus = PanoUserAudio_Unmute;
        [self notifyRefresh:user];
    });
}

- (void)onUserAudioStop:(UInt64)userId {
    dispatch_async(dispatch_get_main_queue(), ^{
        PanoUserInfo *user = [self findUserWithId:userId];
        user.audioStatus = PanoUserAudio_None;
        [self notifyRefresh:user];
    });
}

- (void)onUserAudioMute:(UInt64)userId {
    dispatch_async(dispatch_get_main_queue(), ^{
        PanoUserInfo *user = [self findUserWithId:userId];
        user.audioStatus = PanoUserAudio_Mute;
        [self notifyRefresh:user];
    });
}

- (void)onUserAudioUnmute:(UInt64)userId {
    dispatch_async(dispatch_get_main_queue(), ^{
        PanoUserInfo *user = [self findUserWithId:userId];
        user.audioStatus = PanoUserAudio_Unmute;
        [self notifyRefresh:user];
    });
}

- (void)onUserAudioLevel:(PanoRtcAudioLevel *)level {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.audioLevels setObject:@(level.level) forKey:@(level.userId)];
    });
    if (self.lastSpeakTime == nil) {
        self.lastSpeakTime = [NSDate new];
    }
    if ([[NSDate new] timeIntervalSinceDate:self.lastSpeakTime]> 1) {
        self.lastSpeakTime = [NSDate new];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self invokeWithAction:@selector(onUserAudioLevelChanged) completion:^(id  _Nonnull del) {
                [del onUserAudioLevelChanged];
            }];
        });
    }
}

@end

@implementation PanoUserService (Extension)

- (PanoUserInfo *)host {
    return [self findUserWithId:PanoClientService.service.hostUserId];
}

- (PanoUserInfo *)me {
    return self.dataSource.firstObject;
}

- (PanoUserInfo *)findUserWithId:(UInt64)userId {
    for (PanoUserInfo * userInfo in self.dataSource) {
        if (userInfo.userId == userId) {
            return userInfo;
        }
    }
    return nil;
}

- (NSArray<PanoUserInfo *> *)allUsers {
    return self.dataSource.copy;
}

- (NSArray<PanoUserInfo *> *)allUserExceptMe {
    if (self.dataSource.count <= 1) {
        return @[];
    }
    return [_dataSource subarrayWithRange:NSMakeRange(1, _dataSource.count-1)];
}

- (BOOL)isHost {
    return PanoClientService.service.config.userMode == PanoAnchor;
}

- (NSUInteger)totalCount {
    return _dataSource.count;
}

#pragma mark -- Extension

- (NSInteger)hashValue:(UInt64)hashId {
    NSInteger value = hashId % self.imagesCount;
    if (![self.images[value] boolValue]) {
        self.images[value] = @(true);
        return value;
    }
    for (NSInteger i=0; i<self.imagesCount; i++) {
        if (![self.images[i] boolValue]) {
            self.images[i] = @(true);
            return i;
        }
    }
    return value;
}

- (UIImage *)avatorImage:(UInt64)userId {
    NSString *imageName = self.avators[@(userId)];
    if (imageName) {
        return [UIImage imageNamed:imageName];
    }
    NSInteger value = [self hashValue:userId];
    NSString *prefix = value > 9 ? @"" : @"0";
    imageName = [NSString stringWithFormat:@"%@%ld",prefix, (long)value];
    [self.avators setObject:imageName forKey:@(userId)];
    return [UIImage imageNamed:imageName];
}

@end


@implementation  PanoUserInfo (Add)

- (BOOL)isActiving{
    NSInteger level = [[PanoClientService.service.userService.audioLevels objectForKey:@(self.userId)] intValue];
    if (level > 500) {
        return true;
    }
    return false;
}

- (UIImage *)avatorImage {
    return [PanoClientService.service.userService avatorImage:self.userId];
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Service/PanoUserService.swift
//
//  PanoUserService.swift
//  PanoAudioRoom
//
//  Created by wenhua.zhang on 2020/10/28.
//

import Foundation



public enum PanoUserAudioStatus : Int {
    
    case close = 0
    
    case unmute = 1
    
    case mute = 2
}

class PanoUser : Equatable {
    
    var userId: UInt64 = 0
    
    var userName: String? = ""
    
    var audioStatus: PanoUserAudioStatus = .close
    
    init(userId: UInt64, userName: String?) {
        self.userId = userId
        self.userName = userName
    }
    
    static func == (lhs: PanoUser, rhs: PanoUser) -> Bool {
        return lhs.userId == rhs.userId
    }
}

extension PanoUser {
    func isActiving() -> Bool {
        if let level = PanoUserService.audioLevels[self.userId]  {
            if level > 500 {
                return true
            }
        }
        return false
    }
    
    func avatorImage() -> UIImage! {
        PanoUserService.avatorImage(userId: self.userId)
    }
}


protocol PanoUserInterface {
    
    func me() -> PanoUser?;
    
    func findUser(userId: UInt64) -> PanoUser?;
    
    func allUsers() -> [PanoUser];
    
    func allUserExceptMe() -> [PanoUser]
    
    func totalCount() -> UInt64;
    
}

protocol PanoUserDelegate  {
    
    func onUserAdded(user: PanoUser);
    
    func onUserRemoved(user: PanoUser);
    
    func onUserPropertyChanged(user: PanoUser?);
    
    func onUserAudioLevelChanged();
}

extension PanoUserDelegate {
    
    func onUserPropertyChanged(user: PanoUser?) { };
    
    func onUserAudioLevelChanged() {};
}

class PanoUserService : PanoBaseService, PanoRtcEngineDelegate
{
    static var audioLevels = [UInt64 : UInt32]()
    
    var lastSpeakTime: Date? = nil
    
    lazy var _dataSource : [PanoUser] = []
    
    typealias ObjectType = PanoUserDelegate
    
    var delegates = MulticastDelegate<PanoUserDelegate>();
    
    override class func service() -> PanoUserService? {
        return PanoServiceManager.service(type: .user) as? PanoUserService
    }
    
    override func uninit() {
        self.delegates.removeAllDelegates()
    }
    
    private func notifyRefresh(user: PanoUser?) {
        guard PanoClientService.joined else { return }
        self.delegates.invokeDelegates { (del) in
            del.onUserPropertyChanged(user: user ?? PanoUser(userId: 0, userName: nil))
        }
    }
    
    //MARK: PanoRtcEngineDelegate
    func onUserJoinIndication(_ userId: UInt64, withName userName: String?) {
        DispatchQueue.main.async {
            let user = PanoUser(userId: userId, userName: userName)
            let clientService = PanoServiceManager.service(type: PanoServiceManager.PanoServiceType.client) as! PanoClientService;
            if userId == clientService.userId {
                self._dataSource.insert(user, at: 0)
            } else {
                self._dataSource.append(user)
            }
            print("onUserJoinIndication \(userId) userName: \(String(describing: userName))")
            self.delegates.invokeDelegates { (del: PanoUserDelegate) in
                del.onUserAdded(user: user)
            }
        }
    }
    
    
    func onUserLeaveIndication(_ userId: UInt64, with reason: PanoUserLeaveReason) {
        DispatchQueue.main.async {
            guard PanoClientService.joined else { return }
            let removeUser = PanoUser(userId: userId, userName: nil)
            
            if let index = self._dataSource.firstIndex(of: removeUser) {
                removeUser.userName = self._dataSource[index].userName
                self._dataSource.remove(at: index)
            }
            
            self.delegates.invokeDelegates { (del) in
                del.onUserRemoved(user: removeUser)
            }
        }
    }
    
    func onUserAudioStart(_ userId: UInt64) {
        DispatchQueue.main.async {
            let user = self.findUser(userId: userId)
            user?.audioStatus = PanoUserAudioStatus.unmute
            self.notifyRefresh(user: user)
        }
    }
    
    func onUserAudioStop(_ userId: UInt64) {
        DispatchQueue.main.async {
            let user = self.findUser(userId: userId)
            user?.audioStatus = PanoUserAudioStatus.close
            self.notifyRefresh(user: user)
        }
    }
    
    func onUserAudioMute(_ userId: UInt64) {
        DispatchQueue.main.async {
            let user = self.findUser(userId: userId)
            user?.audioStatus = PanoUserAudioStatus.mute
            self.notifyRefresh(user: user)
        }
    }
    
    func onUserAudioUnmute(_ userId: UInt64) {
        DispatchQueue.main.async {
            let user = self.findUser(userId: userId)
            user?.audioStatus = PanoUserAudioStatus.unmute
            self.notifyRefresh(user: user)
        }
    }
    
    func onUserAudioLevel(_ level: PanoRtcAudioLevel) {
        DispatchQueue.main.async {
            PanoUserService.audioLevels[level.userId] = level.level
        }
        if lastSpeakTime == nil {
            lastSpeakTime = Date()
        }
        if Date().timeIntervalSince(lastSpeakTime!) > 1 {
            lastSpeakTime = Date()
            DispatchQueue.main.async {
                self.delegates.invokeDelegates { (del) in
                    del.onUserAudioLevelChanged();
                }
            }
        }
    }
}

extension PanoUserService {

    var host: PanoUser? {
        get {
            findUser(userId: PanoClientService.hostUserId)
        }
    }
    
    func me() -> PanoUser? {
        return _dataSource.first;
    }
    
    func findUser(userId: UInt64) -> PanoUser? {
        for item in _dataSource {
            if item.userId == userId {
                return item
            }
        }
        return nil
    }
    
    func allUsers() -> [PanoUser] {
        return _dataSource
    }
    
    func allUserExceptMe() -> [PanoUser] {
        if self.totalCount() <= 1 {
            return []
        } else {
            let temp =  Array(_dataSource[1..._dataSource.count-1])
            return temp
        }
    }
    
    func totalCount() -> UInt64 {
        return UInt64(_dataSource.count)
    }
    
    static func isHost() -> Bool {
        return PanoClientService.service().config?.userMode == PanoUserMode.anchor
    }
    
    func addDelegate(delegate: PanoUserDelegate) {
        delegates.addDelegate(delegate)
    }
    
    func removeDelegate(delegate: PanoUserDelegate) {
        delegates.removeDelegate(delegate)
    }
}

extension PanoUserService {
    static var avators = [UInt64 : String]()
    static let imagesCount : Int = 22
    static var images = [Bool](repeating: false, count: imagesCount)
    static var isFull = false;
    class func avatorImage(userId: UInt64) -> UIImage! {
        if let imageName = avators[userId] {
            return UIImage(named: imageName)
        }
        
        func hashValue(id: UInt64) -> Int {
            let value = Int(id % UInt64(imagesCount))
            if images[value] == false  {
                images[value] = true
                return value
            }
            for index in 0..<imagesCount {
                if images[index] == false {
                    images[index] = true
                    return index;
                }
            }
            return value
        }
        let value = hashValue(id: userId)
        let prefix = value > 9 ? "" : "0"
        let name = prefix +  String(value)
        avators[userId] = name
        let image = UIImage(named: name)
        return image
    }
}
